src/sentry/integrations/source_code_management/search.py

```diff
@@ -0,0 +1,103 @@
+
+from django.db.models import Q
+from rest_framework import serializers
+from rest_framework.exceptions import NotFound
+from rest_framework.request import Request
+from rest_framework.response import Response
+
+from sentry.api.api_owners import ApiOwner
+from sentry.api.api_publish_status import ApiPublishStatus
+from sentry.api.base import control_silo_endpoint
+from sentry.integrations.api.bases.integration import IntegrationEndpoint
+from sentry.integrations.models.integration import Integration
+from sentry.integrations.source_code_management.issues import SourceCodeIssueIntegration
+from sentry.organizations.services.organization import RpcOrganization
+
+T = TypeVar("T", bound=SourceCodeIssueIntegration)
+
+
+class SourceCodeSearchSerializer(serializers.Serializer):
+    field = serializers.CharField(required=True)
+    query = serializers.CharField(required=True)
+
+
+@control_silo_endpoint
+class SourceCodeSearchEndpoint(IntegrationEndpoint, Generic[T], ABC):
+    owner = ApiOwner.ECOSYSTEM
+    publish_status = {
+        "GET": ApiPublishStatus.PRIVATE,
+    }
+
+    @property
+    def issue_field(self) -> str:
+        return "externalIssue"
+
+    # not used in VSTS
+    @property
+    def repository_field(self) -> str:
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def integration_provider(self) -> str | None:
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def installation_class(
+        self,
+    ) -> type[T]:
+        raise NotImplementedError
+
+    @abstractmethod
+    def handle_search_issues(self, installation: T, query: str, repo: str) -> Response:
+        raise NotImplementedError
+
+    # not used in VSTS
+    def handle_search_repositories(
+        self, integration: Integration, installation: T, query: str
+    ) -> Response:
+        raise NotImplementedError
+
+    def get(
+        self, request: Request, organization: RpcOrganization, integration_id: int, **kwds: Any
+    ) -> Response:
+        integration_query = Q(
+            organizationintegration__organization_id=organization.id, id=integration_id
+        )
+
+        if self.integration_provider:
+            integration_query &= Q(provider=self.integration_provider)
+        try:
+            integration: Integration = Integration.objects.get(integration_query)
+        except Integration.DoesNotExist:
+            return Response(status=404)
+
+        serializer = SourceCodeSearchSerializer(data=request.query_params)
+        if not serializer.is_valid():
+            return self.respond(serializer.errors, status=400)
+
+        field = serializer.validated_data["field"]
+        query = serializer.validated_data["query"]
+
+        installation = integration.get_installation(organization.id)
+        if not isinstance(installation, self.installation_class):
+            integration_provider = (
+                self.integration_provider if self.integration_provider else "github"
```

Comment:
How could this scenario happen? Seems weird that we'd want to default this to github in the base endpoint class.

Reply:
ah i have self.installation_class as an abstract property anyway, so it must be defined. let me remove this

Reply:
i think i wanted `self.integration_provider` to be defined to raise the `NotFound`, but it would be better to error that the abstract property doesn't exist before doing that. thanks!